packages/beacon-node/src/chain/stateCache/stateContextCheckpointsCache.ts

```diff
@@ -1,43 +1,211 @@
+import path from "node:path";
+import fs from "node:fs";
 import {toHexString} from "@chainsafe/ssz";
 import {phase0, Epoch, RootHex} from "@lodestar/types";
-import {CachedBeaconStateAllForks} from "@lodestar/state-transition";
-import {MapDef} from "@lodestar/utils";
+import {CachedBeaconStateAllForks, computeStartSlotAtEpoch} from "@lodestar/state-transition";
+import {Logger, MapDef, ensureDir, removeFile, writeIfNotExist} from "@lodestar/utils";
 import {routes} from "@lodestar/api";
+import {loadCachedBeaconState} from "@lodestar/state-transition";
 import {Metrics} from "../../metrics/index.js";
+import {LinkedList} from "../../util/array.js";
+import {IClock} from "../../util/clock.js";
+import {ShufflingCache} from "../shufflingCache.js";
 import {MapTracker} from "./mapMetrics.js";
 
 export type CheckpointHex = {epoch: Epoch; rootHex: RootHex};
-const MAX_EPOCHS = 10;
+
+// Make this generic to support testing
+export type PersistentApis = {
+  writeIfNotExist: (filepath: string, bytes: Uint8Array) => Promise<boolean>;
+  removeFile: (path: string) => Promise<boolean>;
+  readFile: (path: string) => Promise<Uint8Array>;
+  ensureDir: (path: string) => Promise<void>;
+};
+
+// Default persistent api for a regular node, use other persistent apis for testing
+const FILE_APIS: PersistentApis = {
+  writeIfNotExist,
+  removeFile,
+  readFile: fs.promises.readFile,
+  ensureDir,
+};
+
+const CHECKPOINT_STATES_FOLDER = "./unfinalized_checkpoint_states";
```

Comment:
at the beginning, I started with writing checkpoint state to `fs` to make it easier to debug, also was worried about db size increase because 2 states persisted per epoch.

Now I switched to persisting 1 state per epoch, make 2 persistent option as `db` and `fs` and make it configurable through cli option. Seems persisting to `db` does not increase its size a lot so will make `db` option default